File: modules/file-transfer/src/main/java/com/gmail/marcosav2010/fth/packet/PacketFileSend.java
package com.gmail.marcosav2010.fth.packet;

import com.gmail.marcosav2010.communicator.packet.Packet;
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketDecoder;
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketEncoder;
import com.gmail.marcosav2010.fth.FileTransferHandler;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.io.IOException;

@NoArgsConstructor
public class PacketFileSend extends Packet {

    /**
     * Data fields + byte arrays + array length
     */
    public static final int MAX_BLOCK_SIZE = Packet.MAX_SIZE - 2 * Integer.BYTES - FileTransferHandler.HASH_SIZE - 2 * Integer.BYTES;

    @Getter
    private int fileID;
    @Getter
    private int pointer;
    @Getter
    private byte[] bytes;
    @Getter
    private byte[] hash;

    public PacketFileSend(int fileID, int pointer, byte[] bytes, byte[] hash) {
        this.fileID = fileID;
        this.pointer = pointer;
        if (bytes.length > MAX_BLOCK_SIZE)
            throw new IllegalArgumentException("Byte block size cannot exceed " + MAX_BLOCK_SIZE + " bytes");
        if (hash.length > FileTransferHandler.HASH_SIZE)
            throw new IllegalArgumentException("Hash size cannot exceed " + FileTransferHandler.HASH_SIZE + " bytes");
        this.bytes = bytes;
        this.hash = hash;
    }

    @Override
    public boolean shouldSendRespose() {
        return false;
    }

    @Override
    protected void encodeContent(PacketEncoder out) throws IOException {
        out.write(fileID);
        out.write(pointer);
        out.write(bytes);
        out.write(hash);
    }

    @Override
    protected void decodeContent(PacketDecoder in) throws IOException {
        fileID = in.readInt();
        pointer = in.readInt();
        bytes = in.readBytes();
        hash = in.readBytes();
    }
}

File: modules/file-transfer/src/main/java/com/gmail/marcosav2010/fth/packet/PacketFileSend.kt
package com.gmail.marcosav2010.fth.packet

import com.gmail.marcosav2010.communicator.packet.Packet
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketDecoder
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketEncoder
import com.gmail.marcosav2010.fth.FileTransferHandler
import kotlin.properties.Delegates

class PacketFileSend() : Packet() {

    companion object {
        /**
         * Data fields + byte arrays + array length
         */
        @JvmField
        val MAX_BLOCK_SIZE = MAX_SIZE - 2 * Integer.BYTES - FileTransferHandler.HASH_SIZE - 2 * Integer.BYTES
    }

    var fileID by Delegates.notNull<Int>()
        private set

    var pointer by Delegates.notNull<Int>()
        private set

    var bytes by Delegates.notNull<ByteArray>()
        private set

    var hash by Delegates.notNull<ByteArray>()
        private set

    constructor(fileID: Int, pointer: Int, bytes: ByteArray, hash: ByteArray) : this() {
        require(bytes.size <= MAX_BLOCK_SIZE) { "Byte block size cannot exceed $MAX_BLOCK_SIZE bytes" }
        require(hash.size <= FileTransferHandler.HASH_SIZE) { "Hash size cannot exceed ${FileTransferHandler.HASH_SIZE} bytes" }
        this.fileID = fileID
        this.pointer = pointer
        this.bytes = bytes
        this.hash = hash
    }

    override fun shouldSendResponse() = false

    override fun encodeContent(out: PacketEncoder) {
        out.write(fileID)
        out.write(pointer)
        out.write(bytes)
        out.write(hash)
    }

    override fun decodeContent(`in`: PacketDecoder) {
        fileID = `in`.readInt()
        pointer = `in`.readInt()
        bytes = `in`.readBytes()
        hash = `in`.readBytes()
    }
}